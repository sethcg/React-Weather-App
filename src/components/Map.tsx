'use client';

import { useEffect, RefObject, useRef } from 'react';
import { Map as LeafletMap, map, Marker, marker, MarkerOptions, IconOptions, icon } from 'leaflet';
import {
  latLng,
  latLngBounds,
  tileLayer,
  MapOptions,
  TileLayerOptions,
  LeafletMouseEvent,
} from 'leaflet';

import Controls from './Controls';

import 'leaflet/dist/leaflet.css';

export default function Map() {
  const url: string = `https://a.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}@2x.png`;

  const options: MapOptions = {
    center: latLng(38.20365531807151, -99.71191406250001),
    minZoom: 3,
    maxZoom: 18,
    zoom: 5,
    zoomSnap: 0.3,
  };
  const tileOptions: TileLayerOptions = {
    tileSize: 512,
    zoomOffset: -1,
    minZoom: 1,
    crossOrigin: true,
    noWrap: true,
  };

  const iconOptions: IconOptions = {
    iconUrl: '/map-marker.svg',
    iconSize: [24, 24],
    iconAnchor: [12, 24],
  };
  const markerOptions: MarkerOptions = { icon: icon(iconOptions) };
  const mapMarker: RefObject<Marker | null> = useRef(marker(latLng(0, 0), markerOptions));

  const mapObject: RefObject<LeafletMap | null> = useRef(null);

  useEffect(() => {
    mapObject.current = map('map', options);

    const bounds = latLngBounds(latLng(-90, -180), latLng(90, 180));
    mapObject.current.setMaxBounds(bounds);

    mapObject.current.on('click', function (event: LeafletMouseEvent) {
      if (mapObject.current === null) return;

      // Remove previous Marker, if necessary
      if (mapMarker.current) {
        mapMarker.current.removeFrom(mapObject.current);
      }

      // Create Marker
      mapMarker.current = marker(event.latlng, markerOptions).addTo(mapObject.current);
    });

    tileLayer(url, tileOptions).addTo(mapObject.current);
  });

  return (
    <div className="grid grid-flow-row-dense grid-cols-3 grid-rows-1 w-full min-h-[calc(100vh-52px)] ps-12 py-6">
      <div className="col-span-2 relative size-full flex justify-start items-center">
          <div id="map" className="size-[calc(100%-4rem)] border-4 border-neutral-800" />
      </div>
      <Controls markerRef={mapMarker} />
    </div>
  );
}
