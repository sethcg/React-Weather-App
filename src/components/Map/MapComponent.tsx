'use client'

import { FunctionComponent, MouseEvent } from 'react'

export const MapComponent: FunctionComponent = () => {
  return (
    <div className="relative min-h-[calc(100%-4rem)] w-full flex-grow">
      <div
        id="map"
        className="border-border pointer-events-auto size-full border-4"
        onContextMenu={(event: MouseEvent) => {
          // Disable left click context menu, so that left click can remove map markers
          event.preventDefault()
        }}
      />
    </div>
  )
}
