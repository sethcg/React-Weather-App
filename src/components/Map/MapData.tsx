'use client'

import { FunctionComponent, useState } from 'react'
import { LatLng, Marker } from 'leaflet'
import { Weather } from '../Weather/Weather'
import { CurrentWeather } from '../Weather/WeatherType'
import { WeatherButton } from '../Controls/WeatherButton'
import { RemoveButton } from '../Controls/RemoveButton'

interface Props {
  mapMarker: Marker
  removeMarker: () => Promise<void>
  zoomToMarker: (latLng: LatLng) => Promise<void>
}

export const MapData: FunctionComponent<Props> = ({ mapMarker, removeMarker, zoomToMarker }) => {
  const lang = 'en'
  const units = 'imperial'

  const [weather, setWeather] = useState<CurrentWeather | null>(null)
  const [previousLatLng, setPreviousLatLng] = useState<LatLng | null>(null)

  const handleProcess = async () => {
    if (!mapMarker) return
    const latLng = mapMarker.getLatLng()

    // Fetch only if the Marker LatLng has changed
    if (!previousLatLng || (previousLatLng.lat !== latLng.lat && previousLatLng.lng !== latLng.lng)) {
      // Fetch weather data
      await fetch(
        `https://api.openweathermap.org/data/2.5/weather?lat=${latLng.lat}&lon=${latLng.lng}&units=${units}&lang=${lang}&appid=${process.env.NEXT_PUBLIC_OPEN_WEATHER_API_KEY}`,
      )
        .then(response => response.json())
        .then(json => {
          // console.log(`[JSON]: ${JSON.stringify(json)}`);
          setWeather(json as CurrentWeather)
        })
        .then(() => {
          // Set previousLatLng to avoid duplicate fetching
          setPreviousLatLng(latLng)
          zoomToMarker(latLng)
        })
        .catch(error => console.error(error))
    }
  }

  const handleRemove = async () => {
    removeMarker()
    setWeather(null)
  }

  return (
    <div className="w-full max-w-[23rem] flex-grow">
      <div className="flex size-full flex-col items-end justify-between border-4 border-neutral-700 py-2">
        <WeatherButton handleProcess={handleProcess} mapMarker={mapMarker} />
        <Weather weatherRef={weather} />
        <RemoveButton handleRemove={handleRemove} />
      </div>
    </div>
  )
}
