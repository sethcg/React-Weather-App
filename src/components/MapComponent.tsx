'use client';

import { FunctionComponent } from 'react';

const MapComponent: FunctionComponent = () => (
  <div className="relative min-h-[calc(100%-4rem)] w-full flex-grow">
    <div id="map" className="size-full border-4 border-neutral-800" />
  </div>
);

export default MapComponent;
