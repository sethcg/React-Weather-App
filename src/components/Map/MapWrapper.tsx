'use client'

import { FunctionComponent, RefObject, useEffect, useRef, useState } from 'react'
import {
  icon,
  IconOptions,
  LatLng,
  latLng,
  latLngBounds,
  LeafletMouseEvent,
  Map as LeafletMap,
  MapOptions,
  marker,
  Marker,
  MarkerOptions,
  tileLayer,
  TileLayerOptions,
} from 'leaflet'

import 'leaflet/dist/leaflet.css'
import { MapData } from './MapData'
import { MapComponent } from './MapComponent'

const url: string = `https://a.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}@2x.png`

const options: MapOptions = {
  center: latLng(38, -99),
  minZoom: 3,
  maxZoom: 14,
  zoom: 5,
  zoomDelta: 0.75,
  zoomSnap: 0.5,
}

const tileOptions: TileLayerOptions = {
  tileSize: 512,
  zoomOffset: -1,
  minZoom: 1,
  crossOrigin: true,
  noWrap: true,
}

const iconOptions: IconOptions = {
  iconUrl: '/marker-icon.png',
  iconSize: [18, 30],
  iconAnchor: [12, 30],
  popupAnchor: [1, -25],
  // shadowUrl: '/marker-shadow.png',
  // shadowSize: [30, 30]
}

const MapWrapper: FunctionComponent = () => {
  const markerOptions: MarkerOptions = { icon: icon(iconOptions) }
  const [mapMarker, setMapMarker] = useState<Marker>(marker(latLng(0, 0), markerOptions))

  const mapObject: RefObject<LeafletMap | null> = useRef(null)

  const createMapMarker = async (latLng: LatLng) => {
    const mapMarkerValue: Marker = marker(latLng, markerOptions)

    // On left click, remove mapMarker
    mapMarkerValue.on('contextmenu', removeMarker)

    return mapMarkerValue
  }

  const addMarker = async (latLng: LatLng) => {
    if (!mapObject.current) return
    const mapMarkerValue: Marker = await createMapMarker(latLng)
    setMapMarker(mapMarkerValue.addTo(mapObject.current))
  }

  const removeMarker = async () => {
    if (!mapObject.current) return
    mapObject.current.getPane('markerPane')?.replaceChildren()
    mapObject.current.getPane('shadowPane')?.replaceChildren()

    const mapMarkerValue: Marker = await createMapMarker(latLng(0, 0))
    setMapMarker(mapMarkerValue)
  }

  const zoomToMarker = async (latLng: LatLng) => {
    if (!mapObject.current) return
    const zoomMin: number = 6
    const zoomMax: number = 11
    const currentZoom: number = mapObject.current.getZoom()
    // clamp the zoom to prevent too large of a jump.
    const zoomLevel: number = Math.min(Math.max(currentZoom, zoomMin), zoomMax)
    mapObject.current.setView(latLng, zoomLevel)
  }

  useEffect(() => {
    mapObject.current?.remove()
    mapObject.current = new LeafletMap('map', options)
    mapObject.current.doubleClickZoom.disable()

    const bounds = latLngBounds(latLng(-90, -180), latLng(90, 180))
    mapObject.current.setMaxBounds(bounds)

    mapObject.current.on('click', function (event: LeafletMouseEvent) {
      removeMarker()
      addMarker(event.latlng)
    })

    tileLayer(url, tileOptions).addTo(mapObject.current)
    // This ensures that mapObject is only created once.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [mapObject])

  return (
    <div className="flex min-h-[calc(100vh-68px)] w-full gap-8 p-8 pt-0">
      <MapComponent />
      <MapData mapMarker={mapMarker} removeMarker={removeMarker} zoomToMarker={zoomToMarker} />
    </div>
  )
}

export default MapWrapper
