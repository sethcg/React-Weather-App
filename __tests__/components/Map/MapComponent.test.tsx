import { render, waitFor } from '@testing-library/react'
import { MapComponent } from '../../../src/components/Map/MapComponent'

describe('MapComponent Snapshot', () => {
  it('Renders MapComponent unchanged', async () => {
    const { container } = await waitFor(() => render(<MapComponent />))

    expect(container).toMatchSnapshot()
  })
})

describe('MapComponent', () => {
  it('Renders the correct div with map ID on the MapComponent', async () => {
    const result = render(<MapComponent />)

    const mapElement = result.container.querySelector('#map')

    expect(mapElement).toBeTruthy()
  })
})
