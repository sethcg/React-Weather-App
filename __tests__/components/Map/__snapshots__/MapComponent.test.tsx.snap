// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapComponent Snapshot Renders MapComponent unchanged 1`] = `
<div>
  <div
    class="relative min-h-[calc(100%-4rem)] w-full flex-grow"
  >
    <div
      class="border-border pointer-events-auto size-full border-4"
      id="map"
    />
  </div>
</div>
`;
